File: app/src/main/java/io/mazur/fit/view/ProgressDialog.java
package io.mazur.fit.view;

import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.support.design.widget.FloatingActionButton;
import android.view.View;
import android.view.Window;
import android.widget.ImageButton;
import android.widget.TextView;

import butterknife.ButterKnife;
import butterknife.InjectView;

import io.mazur.fit.R;
import io.mazur.fit.model.Exercise;
import io.mazur.fit.model.SectionMode;
import io.mazur.fit.stream.RoutineStream;

public class ProgressDialog {
    @InjectView(R.id.level_title_text_view) TextView mLevelTitleTextView;
    @InjectView(R.id.level_previous_button) ImageButton mLevelPreviousButton;
    @InjectView(R.id.level_next_button) ImageButton mLevelNextButton;
    @InjectView(R.id.level_progress_bar) CircularProgressBar mLevelProgressBar;
    @InjectView(R.id.level_text_view) TextView mLevelTextView;
    @InjectView(R.id.level_confirm_button) FloatingActionButton mLevelConfirmButton;

    private Dialog mDialog;

    private Exercise mExercise;

    private int mAvailableLevels;
    private int mChosenLevel = 0;

    public ProgressDialog(Context context, Exercise exercise) {
        mExercise = exercise;
        mAvailableLevels = mExercise.getSection().getAvailableLevels();
        mChosenLevel = mExercise.getSection().getCurrentLevel();

        mDialog = new Dialog(context);
        mDialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        mDialog.setContentView(R.layout.view_dialog_level);
        mDialog.setCanceledOnTouchOutside(true);
    }

    public void show() {
        ButterKnife.inject(this, mDialog);

        mLevelPreviousButton.setOnClickListener(v -> {
            mChosenLevel = mChosenLevel - 1;
            updateDialog();
        });

        mLevelNextButton.setOnClickListener(v -> {
            mChosenLevel = mChosenLevel + 1;
            updateDialog();
        });

        mLevelConfirmButton.setOnClickListener(v -> {
            Exercise chosenExercise = mExercise.getSection().getExercises().get(mChosenLevel);

            RoutineStream.getInstance().setLevel(chosenExercise, mChosenLevel);

            mDialog.dismiss();
        });

        mLevelProgressBar.setWheelSize(12);
        mLevelProgressBar.setProgressColor(Color.parseColor("#009688")); // 00453E
        mLevelProgressBar.setProgressBackgroundColor(Color.parseColor("#00453E"));

        updateDialog();

        mDialog.show();
    }

    private void updateDialog() {
        Exercise chosenExercise = mExercise.getSection().getExercises().get(mChosenLevel);

        if(mExercise.getSection().getSectionMode() == SectionMode.LEVELS) {
            mLevelTitleTextView.setText(chosenExercise.getTitle());
            mLevelTextView.setText(chosenExercise.getLevel());
        } else {
            mLevelTitleTextView.setText(chosenExercise.getTitle());
            mLevelTextView.setText("Pick One");
        }

        if(mChosenLevel == 0) {
            mLevelPreviousButton.setVisibility(View.INVISIBLE);
        } else {
            mLevelPreviousButton.setVisibility(View.VISIBLE);
        }

        if(mChosenLevel >= (mAvailableLevels - 1)) {
            mLevelNextButton.setVisibility(View.INVISIBLE);
        } else {
            mLevelNextButton.setVisibility(View.VISIBLE);
        }

        updateProgressBar();
    }

    private void updateProgressBar() {
        mLevelProgressBar.setProgress((1f / mAvailableLevels) * (mChosenLevel + 1));
    }
}

File: app/src/main/kotlin/com/bodyweight/fitness/dialog/ProgressDialog.kt
package com.bodyweight.fitness.dialog

import android.app.Dialog
import android.support.design.widget.BottomSheetBehavior
import android.support.design.widget.BottomSheetDialogFragment
import android.support.design.widget.CoordinatorLayout
import android.view.View

import com.bodyweight.fitness.*
import com.bodyweight.fitness.model.*
import com.bodyweight.fitness.repository.Repository
import com.bodyweight.fitness.stream.RoutineStream

import kotlinx.android.synthetic.main.view_dialog_progress.view.*

import kotlin.properties.Delegates

class ProgressDialog : BottomSheetDialogFragment() {
    private val bottomSheetCallback = object : BottomSheetBehavior.BottomSheetCallback() {
        override fun onStateChanged(bottomSheet: View, newState: Int) {
            if (newState == BottomSheetBehavior.STATE_HIDDEN) {
                dismiss()
            }
        }

        override fun onSlide(bottomSheet: View, slideOffset: Float) {}
    }

    private val exercise: Exercise = RoutineStream.exercise
    private val availableLevels: Int = exercise.section!!.availableLevels
    private var chosenLevel: Int = exercise.section!!.currentLevel
    private var layout: View by Delegates.notNull()

    override fun setupDialog(dialog: Dialog?, style: Int) {
        super.setupDialog(dialog, style)

        layout = View.inflate(context, R.layout.view_dialog_progress, null)
        dialog?.setContentView(layout)

        val params = ((layout.parent as View).layoutParams as CoordinatorLayout.LayoutParams)
        val behavior = params.behavior

        if (behavior is BottomSheetBehavior) {
            behavior.peekHeight = 400.toPx(context)
            behavior.setBottomSheetCallback(bottomSheetCallback)
        }

        layout.level_previous_button.setOnClickListener {
            chosenLevel -= 1
            updateDialog();
        }

        layout.level_next_button.setOnClickListener {
            chosenLevel += 1
            updateDialog();
        }

        layout.level_confirm_button.setOnClickListener {
            val chosenExercise = exercise.section!!.exercises[chosenLevel]

            RoutineStream.setLevel(chosenExercise, chosenLevel)

            if (Repository.repositoryRoutineForTodayExists()) {
                val repositoryRoutine = Repository.repositoryRoutineForToday

                Repository.realm.executeTransaction {
                    repositoryRoutine.exercises.filter { it.exerciseId == exercise.exerciseId }.firstOrNull()?.let {
                        it.visible == false
                    }

                    repositoryRoutine.exercises.filter { it.exerciseId == chosenExercise.exerciseId }.firstOrNull()?.let {
                        it.visible == true
                    }
                }
            }

            dialog?.dismiss()
        }

        layout.level_progress_bar.setWheelSize(12)
        layout.level_progress_bar.setProgressColor(primary())
        layout.level_progress_bar.setProgressBackgroundColor(primaryDark());

        updateDialog()
    }

    fun updateDialog() {
        val chosenExercise = exercise.section!!.exercises[chosenLevel]

        layout.toolbar.title = chosenExercise.title
        layout.toolbar.subtitle = chosenExercise.description

        if (exercise.section!!.sectionMode == SectionMode.Levels) {
            layout.level_text_view.text = chosenExercise.level
        } else {
            layout.level_text_view.text = "Pick One"
        }

        if (chosenLevel == 0) {
            layout.level_previous_button.setInvisible()
        } else {
            layout.level_previous_button.setVisible()
        }

        if (chosenLevel >= (availableLevels - 1)) {
            layout.level_next_button.setInvisible()
        } else {
            layout.level_next_button.setVisible()
        }

        layout.level_progress_bar.setProgress((1f / availableLevels) * (chosenLevel + 1));
    }
}